src/fdb5/toc/TocHandler.cc

```diff
@@ -1035,7 +1035,7 @@ std::vector<Index> TocHandler::loadIndexes(bool sorted,
             s >> offset;
             s >> type;
             LOG_DEBUG(debug, LibFdb5) << "TocRecord TOC_INDEX " << path << " - " << offset << std::endl;
-            indexes.push_back( new TocIndex(s, r->header_.serialisationVersion_, currentDirectory(),
+            indexes.push_back( new TocIndex(s, *(dynamic_cast<const TocCatalogue*>(this)), r->header_.serialisationVersion_, currentDirectory(),
```

Comment:
I don't like this at all.

The purpose of dynamic_cast is that it uses RTTI to check that the cast is valid - and returns null if not. Here we are assuming that the return value is non-null without checking - either negating the point (which is that it should be used where we cannot assume this), or creating a possible segfault location.

One alternative would be to use dynamic_cast<const TocCatalogue&> which would throw an exception in the case that this was wrong.

Another would be to use a static_cast. Or (at the top of the function) we could ASSERT that the dynamic cast is non-null, and then use the returned value elsewhere in the function.


But this is telling us that something is not right with the type hierarchy. Either this function ought to move to TocCatalogue, or it ought to accept an argument of type "const Catalogue&". Probably the latter.

This code is smelly.